File: Lista_5/Time.py
class Time:
    def __init__(self, hora, min, seg):
        if self.__validTime(hora, min, seg):
            self.__hora = hora
            self.__min = min
            self.__seg = seg
        else:
            print("Hora inválida, inicializando com 0")
            self.__hora = 0
            self.__min = 0
            self.__seg = 0

    @property
    def hora(self):
        return self.__hora
    
    @property
    def min(self):
        return self.__min   
    
    @property
    def seg(self):
        return self.__seg
    

    @hora.setter
    def hora(self, hora):
        print("Inside a setter")
        if hora < 0 or hora > 23:
            print("Hora inválida, inicializando com 0")
            self.__hora = 0
        self.__hora = hora

    @min.setter
    def min(self, min):
        print("Inside a setter")
        if min < 0 or min > 59:
            print("Minuto inválido, inicializando com 0")
            self.__min = 0
        self.__min = min
    
    @seg.setter
    def seg(self, seg):
        print("Inside a setter")
        if seg < 0 or seg > 59:
            print("Segundo inválido, inicializando com 0")
            self.__seg = 0
        self.__seg = seg

    def mostra_hora(self):
        return f'{self.__hora}:{self.__min}:{self.__seg}'
    
    def isAm(self):
        return self.__hora < 12
    
    def cron(self, outra_hora):
        hora_cron = outra_hora.hora*3600 - self.__hora*3600
        min_cron = outra_hora.min*60 - self.__min*60
        seg_cron = (outra_hora.seg - self.__seg) + min_cron + hora_cron

        if self.__hora < outra_hora.hora:
            return seg_cron
        else: 
            return seg_cron + 86400
    
    def add_seconds(self, seg):
        self.__seg += seg
        if self.__seg > 59:
            self.__min += self.__seg // 60
            self.__seg = self.__seg % 60
        if self.__min > 59:
            self.__hora += self.__min // 60
            self.__min = self.__min % 60
        if self.__hora > 23:
            self.__hora = self.__hora % 24
    
    def __validTime(self, hora, min, seg):
        return hora >= 0 and hora <= 23 and min >= 0 and min <= 59 and seg >= 0 and seg <= 59
  
def main():
    hour = Time(21, 28, 25);
    #hour.hora = 21
    #hour.min = 28
    #hour.seg = 25
    print(hour.mostra_hora())
    print(hour.isAm())

    hour.add_seconds(43200)
    print(hour.mostra_hora())

    hour2 = Time(7, 28, 25)
    #hour2.hora = 7
    #hour2.min = 28
    #hour2.seg = 25

    print(f"Diferença de {hour.cron(hour2)} segundos entre hour e hour2")

    print(f"{hour2.hora}:{hour2.min}:{hour2.seg}")


main()

File: Lista_6/Time.java
class Time{
    private int hora, min, seg;
    
    public static void main(String[] args) {
        Time time = new Time(23, 59, 59);
        Time time2 = new Time(time);
        Time time3 = new Time(19);
        Time time4 = new Time(19, 20);

        System.out.println("Time: " + time.mostraHora());
        System.out.println("Time2: " + time2.mostraHora());

        System.out.println("Time é AM? " + time.isAm());
        System.out.println("Time2 é AM? " + time2.isAm());

        System.out.println("Time2 é maior que time? " + time2.cron(time));
        System.out.println("Time é maior que time2? " + time.cron(time2));
        System.out.println("Time4 é maior que time3? " + time4.cron(time3));
        

        time.addTime(time3);
        System.out.println("Time: " + time.mostraHora());
        time2.addTime(20,40);
        System.out.println("Time: " + time2.mostraHora());
        time3.addTime(12, 10, 30);
        System.out.println("Time: " + time3.mostraHora());
    }

    Time(int hora, int min, int seg){
        if(validateTima(hora, min, seg)){
            this.hora = hora;
            this.min = min;
            this.seg = seg;
        }else{
            this.hora = 0;
            this.min = 0;
            this.seg = 0;
        }
    }

    Time(Time obj){
        this(obj.hora, obj.min, obj.seg);
    }
    
    Time(int hora){
        this(hora,0,0);
    }
    
    Time(int hora, int min){
        this(hora, min, 0);
    }
    
    Time(){
        this(0,0,0);
    }

    public int getHora() {
        return this.hora;
    }
    
    public int getMin() {
        return this.min;
    }
    
    public int getSeg() {
        return seg;
    }

    public void setHora(int hora) {
        if(validateTima(hora, 0, 0))
            this.hora = hora;
    }
    
    public void setMin(int min) {
        if (validateTima(0,min ,0))
            this.min = min;
    }
    
    public void setSeg(int seg) {
        if (validateTima(0,0, seg))
            this.seg = seg;
    }

    public String mostraHora(){
        return (this.hora + ":" + this.min + ":" + this.seg);
    }

    public boolean isAm(){
        return this.hora < 12;
    }

    public int cron(Time outraHora){
        int horaCron = outraHora.hora * 3600 - this.hora*3600;
        int minCron = outraHora.min * 60 - this.hora * 60;
        int segCron = outraHora.seg - this.seg;

        if(this.hora < outraHora.hora)
            return segCron;
        else
            return segCron + 86400;
    }

    public void addTime(int hora, int min, int seg){
        if(validateTima(hora, min, seg)){
            this.hora += hora;
            this.min += min;
            this.seg += seg;
        }
        if(this.seg > 59){
            this.seg -= 60;
            this.min++;
        }
        if(this.min > 59){
            this.min -= 60;
            this.hora++;
        }
        if(this.hora > 23){
            this.hora -= 24;
        }
    }
    
    public void addTime(int min, int seg){
        addTime(0, min, seg);
    }
    
    public void addTime(int seg){
        addTime(0,0, seg);
    }
    
    public void addTime(Time obj){
        addTime(obj.hora, obj.min, obj.seg);
    }

    private boolean validateTima(int hora, int min, int seg){
        return ((hora >= 0 && hora <= 23) && (min >= 0 && min <= 59) && (seg >= 0 && seg <= 59));
    }
}